File: frontend/js/views/roth_conversion_ladder.js
// Roth Conversion Ladder — early-retirement bridge. Convert $X/yr from a
// Traditional IRA to Roth; each conversion seasons 5 years before its principal
// is penalty-free. Computation runs server-side via /calc/roth-conversion-ladder
// (traderview-core::roth_conversion_ladder) — a faithful port of the former
// client-side bracket/ladder math, Python-pinned and unit-tested. Class-based
// styling for release-WebKit correctness.

import { api } from '../api.js';
import { applyUiI18n } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const fmt = (n, d) => (n == null || !Number.isFinite(Number(n)) ? '—' : Number(n).toLocaleString(undefined, { minimumFractionDigits: d, maximumFractionDigits: d }));
const VIEW = 'roth-conversion-ladder';
let lastReport = null;
let lastBody = null;

export async function renderRothConversionLadder(mount, _state) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.roth_conversion_ladder.title">// ROTH CONVERSION LADDER</span></h1>
        <p class="muted small" data-i18n-html="view.roth_conversion_ladder.intro">
            Early-retirement bridge. Each $X moved from Traditional → Roth IRA
            becomes principal that can be withdrawn tax- and penalty-free
            after a <strong>5-year season-out period</strong> (per conversion).
            Stagger conversions to build a Roth "ladder" you can draw from
            before age 59½ without the 10% early-withdrawal penalty.
            Conversions count as ordinary income — size them to fill the
            <strong>12% or 22% brackets</strong> (whichever is your seam).
        </p>
        <div class="chart-panel">
            <div class="re-grid">
                <label><span class="muted small">Filing status</span>
                    <select id="rcl-status">
                        <option value="single">Single</option>
                        <option value="mfj" selected>MFJ</option>
                        <option value="mfs">MFS</option>
                        <option value="hoh">HoH</option>
                    </select></label>
                <label><span class="muted small">Other ordinary income / yr $</span>
                    <input type="number" id="rcl-other" step="500" min="0" value="0"></label>
                <label><span class="muted small">Annual conversion $</span>
                    <input type="number" id="rcl-amount" step="1000" min="0" value="60000"></label>
                <label><span class="muted small">Years to ladder</span>
                    <input type="number" id="rcl-years" step="1" min="1" max="30" value="10"></label>
                <label><span class="muted small">Current age</span>
                    <input type="number" id="rcl-age" step="1" min="18" max="65" value="45"></label>
                <label><span class="muted small">Trad IRA balance $</span>
                    <input type="number" id="rcl-balance" step="10000" min="0" value="1200000"></label>
            </div>
            <button class="btn btn-sm primary" id="rcl-run">⚡ Build ladder</button>
            <div id="rcl-tools" class="ce-toolbar"></div>
            <div id="rcl-result" class="re-result"></div>
        </div>
    `;
    applyUiI18n(mount);

    const num = (id) => parseFloat(mount.querySelector(id).value) || 0;
    const readBody = () => ({
        filing_status: mount.querySelector('#rcl-status').value,
        other_income_usd: num('#rcl-other'),
        annual_conversion_usd: num('#rcl-amount'),
        years: Math.max(1, Math.min(30, parseInt(mount.querySelector('#rcl-years').value, 10) || 1)),
        current_age: parseInt(mount.querySelector('#rcl-age').value, 10) || 45,
        traditional_balance_usd: num('#rcl-balance'),
    });
    const compute = async () => {
        const body = readBody();
        if (body.annual_conversion_usd <= 0) {
            mount.querySelector('#rcl-result').innerHTML = `<p class="muted">Conversion amount &gt; 0 required.</p>`;
            lastReport = null; return;
        }
        try {
            const r = await api.calcRothConversionLadder(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            await renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || 'Could not build the ladder.', { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#rcl-tools'), {
        viewId: VIEW, link: false, filename: 'roth-conversion-ladder.csv',
        getRows: () => reportRows(lastReport),
    });
    mount.querySelectorAll('#rcl-status, #rcl-other, #rcl-amount, #rcl-years, #rcl-age, #rcl-balance').forEach(el => {
        el.addEventListener('input', debounce(compute, 250));
    });
    mount.querySelector('#rcl-run').addEventListener('click', compute);
    compute();
}

function reportRows(r) {
    if (!r || !r.valid) return [];
    const rows = [['convert_year', 'age', 'converted_usd', 'tax_usd', 'effective_pct', 'marginal_pct', 'access_age', 'trad_balance_after_usd']];
    for (const x of r.rows) rows.push([x.convert_year, x.age, x.converted_usd, x.tax_on_conversion_usd, x.effective_rate_pct, x.marginal_rate_pct, x.access_age, x.traditional_balance_after_usd]);
    return rows;
}

async function renderResult(mount, r, body, tok) {
    const result = mount.querySelector('#rcl-result');
    if (!r.valid) { result.innerHTML = ''; return; }
    // Line chart: average effective conversion-tax rate as the annual amount sweeps
    // 0 → 3× current (rises as conversions spill into higher brackets).
    const amt = body.annual_conversion_usd || 60000;
    const xs = enh.linspace(amt * 0.25, amt * 3, 12);
    const pts = await Promise.all(xs.map(async (a) => {
        const rr = await api.calcRothConversionLadder({ ...body, annual_conversion_usd: a });
        return { x: a / 1000, y: rr && rr.valid ? rr.avg_effective_rate_pct : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'conversion $k', ylabel: 'avg eff %' });
    result.innerHTML = `
        <div class="cards">
            <div class="card"><div class="label">Total converted</div><div class="value pos">$${fmt(r.total_converted_usd, 0)}</div></div>
            <div class="card"><div class="label">Total fed tax cost</div><div class="value neg">$${fmt(r.total_tax_usd, 0)}</div></div>
            <div class="card"><div class="label">Avg effective rate</div><div class="value">${fmt(r.avg_effective_rate_pct, 2)}%</div></div>
            <div class="card"><div class="label">Trad IRA remaining</div><div class="value">$${fmt(r.traditional_remaining_usd, 0)}</div></div>
            <div class="card"><div class="label">First access at age</div><div class="value">${r.first_access_age ?? '—'}</div></div>
        </div>
        ${chart}
        <table class="trades" data-table-key="rcl-rows">
            <thead><tr>
                <th>Year</th><th>Age</th><th>Convert $</th><th>Tax $</th><th>Effective</th><th>Marginal</th><th>Access age</th><th>Trad balance after</th>
            </tr></thead>
            <tbody>${r.rows.map(x => `<tr>
                <td>${x.convert_year}</td>
                <td>${x.age}</td>
                <td>$${fmt(x.converted_usd, 0)}</td>
                <td class="neg">$${fmt(x.tax_on_conversion_usd, 0)}</td>
                <td>${fmt(x.effective_rate_pct, 2)}%</td>
                <td>${fmt(x.marginal_rate_pct, 0)}%</td>
                <td><strong>${x.access_age}</strong> <span class="muted small">(${x.access_year})</span></td>
                <td class="muted">$${fmt(x.traditional_balance_after_usd, 0)}</td>
            </tr>`).join('')}</tbody>
        </table>
    `;
}

File: crates/traderview-core/src/roth_conversion_ladder.rs
//! Roth conversion ladder — an early-retirement bridge. Each year a fixed amount
//! is converted from a traditional IRA to Roth; the conversion is ordinary income
//! taxed at the marginal bracket above any other income, and the converted
//! principal becomes penalty-free to withdraw after a 5-year season-out. This
//! walks the ladder year by year: tax cost per conversion, effective/marginal
//! rate, the age each year's conversion becomes accessible, and the draining
//! traditional balance. TY2025 brackets (Rev. Proc. 2024-40). Faithful port of
//! the former client-side calculator. Pure compute, not advice.

use serde::{Deserialize, Serialize};

fn brackets(status: &str) -> &'static [(f64, f64)] {
    match status {
        "mfj" => &[(0.0, 0.10), (23_850.0, 0.12), (96_950.0, 0.22), (206_700.0, 0.24), (394_600.0, 0.32), (501_050.0, 0.35), (751_600.0, 0.37)],
        "mfs" => &[(0.0, 0.10), (11_925.0, 0.12), (48_475.0, 0.22), (103_350.0, 0.24), (197_300.0, 0.32), (250_525.0, 0.35), (375_800.0, 0.37)],
        "hoh" => &[(0.0, 0.10), (17_000.0, 0.12), (64_850.0, 0.22), (103_350.0, 0.24), (197_300.0, 0.32), (250_500.0, 0.35), (626_350.0, 0.37)],
        _ => &[(0.0, 0.10), (11_925.0, 0.12), (48_475.0, 0.22), (103_350.0, 0.24), (197_300.0, 0.32), (250_525.0, 0.35), (626_350.0, 0.37)],
    }
}

fn std_deduction(status: &str) -> f64 {
    match status {
        "mfj" => 30_000.0,
        "hoh" => 22_500.0,
        _ => 15_000.0,
    }
}

const BASE_YEAR: i32 = 2026;
const SEASON_YEARS: i32 = 5;

#[derive(Debug, Clone, Deserialize)]
pub struct LadderInput {
    pub filing_status: String,
    #[serde(default)]
    pub other_income_usd: f64,
    pub annual_conversion_usd: f64,
    pub years: u32,
    pub current_age: i32,
    pub traditional_balance_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct LadderRow {
    pub convert_year: i32,
    pub age: i32,
    pub converted_usd: f64,
    pub tax_on_conversion_usd: f64,
    pub effective_rate_pct: f64,
    pub marginal_rate_pct: f64,
    pub access_age: i32,
    pub access_year: i32,
    pub traditional_balance_after_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct LadderReport {
    pub total_converted_usd: f64,
    pub total_tax_usd: f64,
    pub avg_effective_rate_pct: f64,
    pub traditional_remaining_usd: f64,
    pub first_access_age: Option<i32>,
    pub rows: Vec<LadderRow>,
    pub valid: bool,
}

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

fn tax_on(brk: &[(f64, f64)], taxable: f64) -> f64 {
    let mut tax = 0.0;
    for (idx, &(from, rate)) in brk.iter().enumerate() {
        if taxable <= from {
            break;
        }
        let to = brk.get(idx + 1).map(|b| b.0);
        let top = to.map_or(taxable, |t| taxable.min(t));
        tax += (top - from).max(0.0) * rate;
    }
    tax
}

fn marginal_at(brk: &[(f64, f64)], taxable: f64) -> f64 {
    let mut m = 0.0;
    for &(from, rate) in brk {
        if taxable >= from {
            m = rate;
        }
    }
    m
}

pub fn generate(i: &LadderInput) -> LadderReport {
    if i.annual_conversion_usd <= 0.0 {
        return LadderReport::default();
    }
    let years = i.years.clamp(1, 30);
    let brk = brackets(&i.filing_status);
    let ded = std_deduction(&i.filing_status);
    let baseline_taxable = (i.other_income_usd - ded).max(0.0);
    let baseline_tax = tax_on(brk, baseline_taxable);

    let mut balance = i.traditional_balance_usd;
    let mut rows = Vec::new();
    let mut total_tax = 0.0;
    let mut total_converted = 0.0;
    for y in 0..years as i32 {
        if balance <= 0.0 {
            break;
        }
        let convert = i.annual_conversion_usd.min(balance);
        balance -= convert;
        let taxable = (i.other_income_usd + convert - ded).max(0.0);
        let tax_on_conv = tax_on(brk, taxable) - baseline_tax;
        let effective = if convert > 0.0 { tax_on_conv / convert } else { 0.0 };
        rows.push(LadderRow {
            convert_year: BASE_YEAR + y,
            age: i.current_age + y,
            converted_usd: round2(convert),
            tax_on_conversion_usd: round2(tax_on_conv),
            effective_rate_pct: round4(effective * 100.0),
            marginal_rate_pct: round4(marginal_at(brk, taxable) * 100.0),
            access_age: i.current_age + y + SEASON_YEARS,
            access_year: BASE_YEAR + y + SEASON_YEARS,
            traditional_balance_after_usd: round2(balance),
        });
        total_tax += tax_on_conv;
        total_converted += convert;
    }

    let avg_eff = if total_converted > 0.0 { total_tax / total_converted } else { 0.0 };
    LadderReport {
        total_converted_usd: round2(total_converted),
        total_tax_usd: round2(total_tax),
        avg_effective_rate_pct: round4(avg_eff * 100.0),
        traditional_remaining_usd: round2(balance),
        first_access_age: rows.first().map(|r| r.access_age),
        rows,
        valid: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> LadderInput {
        LadderInput {
            filing_status: "mfj".into(),
            other_income_usd: 0.0,
            annual_conversion_usd: 60_000.0,
            years: 10,
            current_age: 45,
            traditional_balance_usd: 1_200_000.0,
        }
    }

    // Pins cross-checked against the original JS compute() in Python.
    #[test]
    fn default_mfj_ladder() {
        let d = generate(&base());
        assert_eq!(d.rows.len(), 10);
        assert!(close(d.rows[0].tax_on_conversion_usd, 3_123.0));
        assert!(close(d.total_converted_usd, 600_000.0));
        assert!(close(d.total_tax_usd, 31_230.0));
        assert!(close(d.avg_effective_rate_pct, 5.205));
        assert!(close(d.traditional_remaining_usd, 600_000.0));
        assert_eq!(d.first_access_age, Some(50));
        assert!(close(d.rows[0].marginal_rate_pct, 12.0));
    }

    #[test]
    fn ladder_stops_when_balance_drained() {
        // Small balance covers only 2 full conversions plus a partial.
        let d = generate(&LadderInput { traditional_balance_usd: 130_000.0, ..base() });
        assert_eq!(d.rows.len(), 3);
        assert!(close(d.rows[2].converted_usd, 10_000.0)); // 130k - 60k - 60k
        assert!(close(d.traditional_remaining_usd, 0.0));
    }

    #[test]
    fn access_age_is_five_years_out() {
        let d = generate(&base());
        for r in &d.rows {
            assert_eq!(r.access_age, r.age + 5);
        }
    }

    #[test]
    fn other_income_raises_conversion_tax() {
        // Other income pushes the conversion into higher brackets → more tax than baseline-zero.
        let d = generate(&LadderInput { other_income_usd: 100_000.0, ..base() });
        assert!(d.rows[0].tax_on_conversion_usd > 3_123.0);
    }

    #[test]
    fn zero_conversion_invalid() {
        let d = generate(&LadderInput { annual_conversion_usd: 0.0, ..base() });
        assert!(!d.valid);
    }
}
